### PR TITLE
Issue #540 - Added a LifeCycleService + Bugfix when recovering from maintenance mode

### PR DESCRIPTION
**Pull request by mathieucarbou**
@alexsnaps, @ljacomet, @lorban : pull request for #540 (service part), with the modifications we talked about.

I used a `StateChangeListener` to hook into the state transitioner because these listeners are triggered after all init / close hooks are executed, which is what we need to be aware of the initialized and closed states I think (?).

Also fixed a missing `succeeded();` call when closing maintenance mode

Added 2 tests for both the issue and the LifeCycle Service.

### CODE REVIEW

**Comment by ljacomet** [comment]
Your branch is based off the wrong commit - please fix

**Comment by mathieucarbou** [comment]
@ljacomet : done.

File: core/src/main/java/org/ehcache/EhcacheManager.java [hunk 1]
```diff
@@ -371,12 +388,17 @@ public void close() {
     
     RuntimeConfiguration<K, V> runtimeConfiguration = new RuntimeConfiguration<K, V>(config, evtService);
     runtimeConfiguration.addCacheConfigurationListener(store.getConfigurationChangeListeners());
+
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/EhcacheManager.java]: Please no unrelated formatting changes

File: core/src/main/java/org/ehcache/EhcacheManager.java [hunk 2]
```diff
@@ -371,12 +388,17 @@ public void close() {
...
     final Ehcache<K, V> ehCache = new Ehcache<K, V>(runtimeConfiguration, store, decorator, evtService,
         useLoaderInAtomics, LoggerFactory.getLogger(Ehcache.class + "-" + alias));
 
     final ManagementRegistry managementRegistry = serviceLocator.getService(ManagementRegistry.class);
     final EhcacheStatsSettings ehcacheStatsSettings = new EhcacheStatsSettings(alias, Collections.<String, Object>singletonMap("Setting", "CacheName"));
 
+    // registers a listener on ehcache status transitioner, when all init / close hoot are done
+    // to provide a way for services to know when an ehcache instance is initialized or closed
+    ehCache.registerListener(lifeCycleManager.createStateChangeListener(ehCache));
+
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/EhcacheManager.java]: Configure IDE to not have trailing spaces like here

File: core/src/main/java/org/ehcache/spi/lifecycle/LifeCycleListener.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * Listens of life cycle change on a particular object
+ * 
+ * @author Mathieu Carbou
+ */
+public interface LifeCycleListener<T> {
+
+  /**
+   * Called when the instance is initialized
+   */
+  void onInitialized(T instance);
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/spi/lifecycle/LifeCycleListener.java]: I feel `afterInitialisation(T instance)` would better indicate when you are called exactly.

File: core/src/main/java/org/ehcache/spi/lifecycle/LifeCycleListener.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Called when the instance is closed
+   */
+  void onClosed(T instance);
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/spi/lifecycle/LifeCycleListener.java]: I feel `afterClosing(T instance)` would better indicate when you are called exactly.

**Comment by mathieucarbou** [comment]
@ljacomet : fixed trailing space, new line, method names, updated javadoc, and also rebased in 1 commit.
The build failed because I forced a push...

**Comment by ljacomet** [comment]
test this please

**Comment by mathieucarbou** [comment]
@ljacomet : you mean only the LifeCycleListenerAdapter right ? Because there is already LifeCycleTest added plus EhcacheManagerTest updated.

**Comment by ljacomet** [comment]
@mathieucarbou Sorry, do not get the question. What are you referring to?

**Comment by mathieucarbou** [comment]
@ljacomet : your comment 2 days ago: "test this please" ;-)

**Comment by mathieucarbou** [comment]
rebased, added test, made sure the listener to clear listener references is only added once.

**Comment by mathieucarbou** [comment]
test this please

**Comment by mathieucarbou** [comment]
test this please

**Comment by mathieucarbou** [comment]
test this please

**Comment by mathieucarbou** [comment]
test this please

**Comment by mathieucarbou** [comment]
will create a pull request with several separated commits
